### PR TITLE
PML for H field

### PR DESCRIPTION
**Pull request by ajnonaka**
This PR implements PML boundary conditions for the H field for simulations with USE_LLG_TRUE.
The goal is to replicate how PML is done for the B field, but use the correct scaling of mu_0 (vacuum for now) in the PML region advancement.

This animation shows the Hx field gracefully exiting the domain using Examples/Tests/Macroscopic_Maxwell/inputs_3d_LLG_noMs

PML

### CODE REVIEW

File: Source/BoundaryConditions/PML.cpp [hunk 1]
```diff
@@ -506,14 +506,26 @@ PML::PML (const BoxArray& grid_ba, const DistributionMapping& /*grid_dm*/,
...
     pml_B_fp[2] = std::make_unique<MultiFab>(amrex::convert( ba,
         WarpX::GetInstance().getBfield_fp(0,2).ixType().toIntVect() ), dm, 2, ngb );
-
+#ifdef WARPX_MAG_LLG
+    pml_H_fp[0] = std::make_unique<MultiFab>(amrex::convert( ba,
+        WarpX::GetInstance().getHfield_fp(0,0).ixType().toIntVect() ), dm, 2, ngb );
+    pml_H_fp[1] = std::make_unique<MultiFab>(amrex::convert( ba,
+        WarpX::GetInstance().getHfield_fp(0,1).ixType().toIntVect() ), dm, 2, ngb );
+    pml_H_fp[2] = std::make_unique<MultiFab>(amrex::convert( ba,
+        WarpX::GetInstance().getHfield_fp(0,2).ixType().toIntVect() ), dm, 2, ngb );
```
Comment (RevathiJambunathan) [inline Source/BoundaryConditions/PML.cpp:515]: Something similar can be done for coarse patch H multifabs as well in the condiitonal` if(cgeom)`

**Review by RevathiJambunathan — CHANGES_REQUESTED** [review]

**Comment by RevathiJambunathan** [comment]
Thank you for this PR, @ajnonaka . I just started reviewing but I thought I might as well post the one comment I had to communicate the definitions for the coarsepatch H multifabs. 
While I continue to review the rest of the PR.

File: Source/BoundaryConditions/PML.cpp (hunk 1 above)
Comment (ajnonaka) [reply Source/BoundaryConditions/PML.cpp:515]: fixed with 482ec751

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveHPML.cpp [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        Real const * const AMREX_RESTRICT coefs_y = m_stencil_coefs_y.dataPtr();
+        int const n_coefs_y = m_stencil_coefs_y.size();
+        Real const * const AMREX_RESTRICT coefs_z = m_stencil_coefs_z.dataPtr();
+        int const n_coefs_z = m_stencil_coefs_z.size();
+
+        // Extract tileboxes for which to loop
+        Box const& tbx  = mfi.tilebox(Hfield[0]->ixType().ixType());
+        Box const& tby  = mfi.tilebox(Hfield[1]->ixType().ixType());
+        Box const& tbz  = mfi.tilebox(Hfield[2]->ixType().ixType());
+
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/EvolveHPML.cpp:92]:
```suggestion
        Box const& tbz  = mfi.tilebox(Hfield[2]->ixType().ixType());
        amrex::Real mu0_inv = 1._rt/PhysConst::mu0
```

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveHPML.cpp [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        // Loop over the cells and update the fields
+        amrex::ParallelFor(tbx, tby, tbz,
+
+            [=] AMREX_GPU_DEVICE (int i, int j, int k){
+                Hx(i, j, k, PMLComp::xz) += dt * (
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/EvolveHPML.cpp]:
```suggestion
                Hx(i, j, k, PMLComp::xz) += mu0_inv * dt * (
```

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveHPML.cpp [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+                    T_Algo::UpwardDz(Ey, coefs_z, n_coefs_z, i, j, k, PMLComp::yx)
+                  + T_Algo::UpwardDz(Ey, coefs_z, n_coefs_z, i, j, k, PMLComp::yy)
+                  + T_Algo::UpwardDz(Ey, coefs_z, n_coefs_z, i, j, k, PMLComp::yz) )
+                    / PhysConst::mu0;
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/EvolveHPML.cpp]:
```suggestion
```

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveHPML.cpp [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        Box const& tbz  = mfi.tilebox(Hfield[2]->ixType().ixType());
+
+        // Loop over the cells and update the fields
+        amrex::ParallelFor(tbx, tby, tbz,
+
+            [=] AMREX_GPU_DEVICE (int i, int j, int k){
+                Hx(i, j, k, PMLComp::xz) += dt * (
+                    T_Algo::UpwardDz(Ey, coefs_z, n_coefs_z, i, j, k, PMLComp::yx)
+                  + T_Algo::UpwardDz(Ey, coefs_z, n_coefs_z, i, j, k, PMLComp::yy)
+                  + T_Algo::UpwardDz(Ey, coefs_z, n_coefs_z, i, j, k, PMLComp::yz) )
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/EvolveHPML.cpp]:
```suggestion
                  + T_Algo::UpwardDz(Ey, coefs_z, n_coefs_z, i, j, k, PMLComp::yz) );
```

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveHPML.cpp [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+                    / PhysConst::mu0;
+                Hx(i, j, k, PMLComp::xy) -= dt * (
+                    T_Algo::UpwardDy(Ez, coefs_y, n_coefs_y, i, j, k, PMLComp::zx)
+                  + T_Algo::UpwardDy(Ez, coefs_y, n_coefs_y, i, j, k, PMLComp::zy)
+                  + T_Algo::UpwardDy(Ez, coefs_y, n_coefs_y, i, j, k, PMLComp::zz) )
+                    / PhysConst::mu0;
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/EvolveHPML.cpp]:
```suggestion
                Hx(i, j, k, PMLComp::xy) -= mu0_inv * dt * (
                    T_Algo::UpwardDy(Ez, coefs_y, n_coefs_y, i, j, k, PMLComp::zx)
                  + T_Algo::UpwardDy(Ez, coefs_y, n_coefs_y, i, j, k, PMLComp::zy)
                  + T_Algo::UpwardDy(Ez, coefs_y, n_coefs_y, i, j, k, PMLComp::zz) );
```

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveHPML.cpp [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+            },
+
+            [=] AMREX_GPU_DEVICE (int i, int j, int k){
+                Hy(i, j, k, PMLComp::yx) += dt * (
+                    T_Algo::UpwardDx(Ez, coefs_x, n_coefs_x, i, j, k, PMLComp::zx)
+                  + T_Algo::UpwardDx(Ez, coefs_x, n_coefs_x, i, j, k, PMLComp::zy)
+                  + T_Algo::UpwardDx(Ez, coefs_x, n_coefs_x, i, j, k, PMLComp::zz) )
+                    / PhysConst::mu0;
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/EvolveHPML.cpp]:
```suggestion
                Hy(i, j, k, PMLComp::yx) += mu0_inv * dt * (
                    T_Algo::UpwardDx(Ez, coefs_x, n_coefs_x, i, j, k, PMLComp::zx)
                  + T_Algo::UpwardDx(Ez, coefs_x, n_coefs_x, i, j, k, PMLComp::zy)
                  + T_Algo::UpwardDx(Ez, coefs_x, n_coefs_x, i, j, k, PMLComp::zz) );
```

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveHPML.cpp [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+                Hy(i, j, k, PMLComp::yz) -= dt * (
+                    T_Algo::UpwardDz(Ex, coefs_z, n_coefs_z, i, j, k, PMLComp::xx)
+                  + T_Algo::UpwardDz(Ex, coefs_z, n_coefs_z, i, j, k, PMLComp::xy)
+                  + T_Algo::UpwardDz(Ex, coefs_z, n_coefs_z, i, j, k, PMLComp::xz) )
+                    / PhysConst::mu0;
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/EvolveHPML.cpp]:
```suggestion
                Hy(i, j, k, PMLComp::yz) -= mu0_inv * dt * (
                    T_Algo::UpwardDz(Ex, coefs_z, n_coefs_z, i, j, k, PMLComp::xx)
                  + T_Algo::UpwardDz(Ex, coefs_z, n_coefs_z, i, j, k, PMLComp::xy)
                  + T_Algo::UpwardDz(Ex, coefs_z, n_coefs_z, i, j, k, PMLComp::xz) );
```

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveHPML.cpp [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+            },
+
+            [=] AMREX_GPU_DEVICE (int i, int j, int k){
+                Hz(i, j, k, PMLComp::zy) += dt * (
+                    T_Algo::UpwardDy(Ex, coefs_y, n_coefs_y, i, j, k, PMLComp::xx)
+                  + T_Algo::UpwardDy(Ex, coefs_y, n_coefs_y, i, j, k, PMLComp::xy)
+                  + T_Algo::UpwardDy(Ex, coefs_y, n_coefs_y, i, j, k, PMLComp::xz) )
+                    / PhysConst::mu0;
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/EvolveHPML.cpp]:
```suggestion
                Hz(i, j, k, PMLComp::zy) += mu0_inv * dt * (
                    T_Algo::UpwardDy(Ex, coefs_y, n_coefs_y, i, j, k, PMLComp::xx)
                  + T_Algo::UpwardDy(Ex, coefs_y, n_coefs_y, i, j, k, PMLComp::xy)
                  + T_Algo::UpwardDy(Ex, coefs_y, n_coefs_y, i, j, k, PMLComp::xz) );
```

File: Source/FieldSolver/FiniteDifferenceSolver/EvolveHPML.cpp [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+                Hz(i, j, k, PMLComp::zx) -= dt * (
+                    T_Algo::UpwardDx(Ey, coefs_x, n_coefs_x, i, j, k, PMLComp::yx)
+                  + T_Algo::UpwardDx(Ey, coefs_x, n_coefs_x, i, j, k, PMLComp::yy)
+                  + T_Algo::UpwardDx(Ey, coefs_x, n_coefs_x, i, j, k, PMLComp::yz) )
+                    / PhysConst::mu0;
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/EvolveHPML.cpp]:
```suggestion
                Hz(i, j, k, PMLComp::zx) -= mu0_inv * dt * (
                    T_Algo::UpwardDx(Ey, coefs_x, n_coefs_x, i, j, k, PMLComp::yx)
                  + T_Algo::UpwardDx(Ey, coefs_x, n_coefs_x, i, j, k, PMLComp::yy)
                  + T_Algo::UpwardDx(Ey, coefs_x, n_coefs_x, i, j, k, PMLComp::yz) );
```

File: Source/FieldSolver/FiniteDifferenceSolver/Make.package [hunk 11]
```diff
@@ -8,6 +8,7 @@ CEXE_sources += MacroscopicEvolveE.cpp
 #ifdef WARPX_MAG_LLG
 CEXE_sources += MacroscopicEvolveHM.cpp
 CEXE_sources += MacroscopicEvolveHM_2nd.cpp
+CEXE_sources += EvolveHPML.cpp
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/FiniteDifferenceSolver/Make.package:11]: We may also need to add this file to the CMakelists file

File: Source/FieldSolver/WarpXPushFieldsEM.cpp [hunk 12]
```diff
@@ -231,13 +231,23 @@ WarpX::EvolveE (int lev, PatchType patch_type, amrex::Real a_dt)
...
+#endif
                 pml[lev]->Getj_fp(), pml[lev]->GetF_fp(),
                 pml[lev]->GetMultiSigmaBox_fp(),
                 a_dt, pml_has_particles );
         } else {
             m_fdtd_solver_cp[lev]->EvolveEPML(
-                pml[lev]->GetE_cp(), pml[lev]->GetB_cp(),
+                pml[lev]->GetE_cp(),
+#ifdef WARPX_MAG_LLG
+                pml[lev]->GetH_fp(),
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/WarpXPushFieldsEM.cpp]:
```suggestion
                pml[lev]->GetH_cp(),
```

File: Source/FieldSolver/WarpXPushFieldsEM.cpp [hunk 13]
```diff
@@ -231,13 +231,23 @@ WarpX::EvolveE (int lev, PatchType patch_type, amrex::Real a_dt)
...
+#else
+                pml[lev]->GetB_fp(),
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/WarpXPushFieldsEM.cpp]:
```suggestion
                pml[lev]->GetB_cp(),
```

File: Source/FieldSolver/WarpXPushFieldsEM.cpp [hunk 14]
```diff
@@ -330,16 +342,28 @@ WarpX::MacroscopicEvolveE (int lev, PatchType patch_type, amrex::Real a_dt) {
...
+#endif                
                 pml[lev]->Getj_fp(), pml[lev]->GetF_fp(),
                 pml[lev]->GetMultiSigmaBox_fp(),
                 a_dt, pml_has_particles );
         } else {
             m_fdtd_solver_cp[lev]->EvolveEPML(
-                pml[lev]->GetE_cp(), pml[lev]->GetB_cp(),
+                pml[lev]->GetE_cp(),
+#ifdef WARPX_MAG_LLG
+                pml[lev]->GetH_fp(),
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/WarpXPushFieldsEM.cpp]:
```suggestion
                pml[lev]->GetH_cp(),
```

File: Source/FieldSolver/WarpXPushFieldsEM.cpp [hunk 15]
```diff
@@ -330,16 +342,28 @@ WarpX::MacroscopicEvolveE (int lev, PatchType patch_type, amrex::Real a_dt) {
...
+#else
+                pml[lev]->GetB_fp(),
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/WarpXPushFieldsEM.cpp]:
```suggestion
                pml[lev]->GetB_cp(),
```

**Review by RevathiJambunathan — COMMENTED** [review]
Thank you Andy for this PR.
A few suggestions and a note to add the file to CMakelists as well. so we can cmake it :)

**Comment by RevathiJambunathan** [comment]
an eol error. probably from the suggestions I added.

File: Source/FieldSolver/FiniteDifferenceSolver/Make.package (hunk 11 above)
Comment (ajnonaka) [reply Source/FieldSolver/FiniteDifferenceSolver/Make.package:11]: Done.

File: Source/BoundaryConditions/PML.H [hunk 16]
```diff
@@ -149,13 +153,24 @@ public:
     void ExchangeF (amrex::MultiFab* F_fp, amrex::MultiFab* F_cp, int do_pml_in_domain);
     void ExchangeF (PatchType patch_type, amrex::MultiFab* Fp, int do_pml_in_domain);
 
+#ifdef WARPX_MAG_LLG
+    void ExchangeH (const std::array<amrex::MultiFab*,3>& H_fp,
+                    const std::array<amrex::MultiFab*,3>& H_cp, int do_pml_in_domain);
+    void ExchangeH (PatchType patch_type,
+                    const std::array<amrex::MultiFab*,3>& Bp, int do_pml_in_domain);
```
Comment (ajnonaka) [inline Source/BoundaryConditions/PML.H]:
```suggestion
                    const std::array<amrex::MultiFab*,3>& Hp, int do_pml_in_domain);
```

**Comment by RevathiJambunathan** [comment]
@ajnonaka thank you for the cool animation and this PR!

**Review by RevathiJambunathan — APPROVED** [review]

**Pull request merged**